File: dailyques/Q1339.py
#Q1339.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def maxProduct(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        MOD = 10**9 + 7

        def sums(node):
            if not node: return 0
            return node.val + sums(node.left) + sums(node.right)

        total = sums(root)
        self.ans = 0

        def dfs(node):
            if not node: return 0
            s = node.val + dfs(node.left) + dfs(node.right)
            self.ans = max(self.ans, s * (total - s))
            return s

        dfs(root)
        return self.ans % MOD

File: dailyques/Q1339.cpp
//Q1339.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long long total = 0, best = 0;
    const long long MOD = 1000000007;

    long long sum(TreeNode* r){
        if(!r) return 0;
        return r->val + sum(r->left) + sum(r->right);
    }
    long long dfs(TreeNode* r){
        if(!r) return 0;
        long long s = r->val + dfs(r->left) + dfs(r->right);
        best = max(best, s * (total - s));
        return s;
    }

    int maxProduct(TreeNode* root) {
        total = sum(root);
        dfs(root);
        return best % MOD;
    }
};
